File: app/src/main/java/com/tasktree/Peg.java
package com.tasktree;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

abstract class Peg {

    List<Ball> balls;

    public Peg( List<Ball> balls ) {
        this.balls = new ArrayList<>( balls );
    }

    public Position getOpenPositions( ) {
        return NullPosition.INSTANCE;
    }

    public abstract Ball getBallAtPosition( Position position );

    public List<Ball> getAllBalls( ) {
        return Collections.unmodifiableList( balls );
    }
}

File: app/src/main/kotlin/com/tasktree/Peg.kt
package com.tasktree

import java.util.*

internal abstract class Peg(val balls: List<Ball>) {

    init {
        require(balls.size == size) {
            "Invalid list of balls given for this peg expected ${size} given ${balls.size}"
        }
    }

    abstract val size: Int;

    abstract fun openPositions(): List<Position>

    abstract fun hasOpenSlotAt(position: Position): Boolean

    abstract fun getBallAtPosition(position: Position): Ball?

    abstract fun moveBallToPosition(ball: Ball, newPosition: Position): Unit

    fun withinRange(newPosition: Position): Boolean {
        return (0..size-1).contains(newPosition.index)
    }

    val allBalls: List<Ball>
        get() = Collections.unmodifiableList(balls)
}


